packages/script/src/script-factory.ts

```diff
@@ -0,0 +1,145 @@
+
+  constructor(bytecode: BytesLike, abi: JsonAbi, walletOrProvider: BaseWalletLocked | Provider) {
+    this.bytecode = bytecode;
+    this.interface = new Interface(abi);
+
+    if (walletOrProvider && 'provider' in walletOrProvider) {
+      this.provider = walletOrProvider.provider;
+      this.wallet = walletOrProvider;
+    } else {
+      this.provider = walletOrProvider;
+      this.wallet = null;
+    }
+
+    this.script = new Script(
+      bytecode,
+      (args: InputValue<void>[]) =>
+        this.interface.encodeFunctionData(FUNCTION_FRAGMENT_NAME, args, 0, true),
+      (scriptResult): Result<TOutput> => {
+        const logs = getDecodedLogs(scriptResult.receipts, this.interface);
+
+        if (scriptResult.returnReceipt.type === ReceiptType.Revert) {
+          logger.throwError('Script Reverted', Logger.errors.CALL_EXCEPTION, logs);
+        }
+
+        if (
+          scriptResult.returnReceipt.type !== ReceiptType.Return &&
+          scriptResult.returnReceipt.type !== ReceiptType.ReturnData
+        ) {
+          logger.throwError(
+            `Script Return Type [${scriptResult.returnReceipt.type}] Invalid`,
+            Logger.errors.CALL_EXCEPTION,
+            {
+              logs,
+              receipt: scriptResult.returnReceipt,
+            }
+          );
+        }
+
+        let value;
+        if (scriptResult.returnReceipt.type === ReceiptType.Return) {
+          value = scriptResult.returnReceipt.val;
+        }
+        if (scriptResult.returnReceipt.type === ReceiptType.ReturnData) {
+          const decoded = this.interface.decodeFunctionResult(
+            FUNCTION_FRAGMENT_NAME,
+            scriptResult.returnReceipt.data
+          );
+          value = (decoded as [TOutput])[0];
+        }
+
+        return {
+          value,
+          logs,
+        };
+      }
+    );
+  }
+
+  async buildScriptTransaction<TData>(
+    data: InputValue<TData>[],
+    scriptOptions?: BuildScriptOptions
+  ): Promise<ScriptTransactionRequest> {
+    const options = {
+      fundTransaction: true,
+      ...scriptOptions,
+    };
+    const request = new ScriptTransactionRequest({
+      gasLimit: MAX_GAS_PER_TX,
+      ...options,
+    });
+    request.setScript(this.script, data as InputValue[]);
+
+    const requiredCoinQuantities: CoinQuantityLike[] = [];
+    if (options.fundTransaction) {
+      requiredCoinQuantities.push(request.calculateFee());
+    }
+
+    if (requiredCoinQuantities.length && this.wallet) {
+      const resources = await this.wallet.getResourcesToSpend(requiredCoinQuantities);
+      request.addResources(resources);
+    }
+
+    return request;
+  }
+
+  async call(
+    data: InputValue<TInput>[],
+    options?: BuildScriptOptions
+  ): Promise<
+    {
+      transactionResult: TransactionResult<'success' | 'failure'>;
+      response: TransactionResponse;
+    } & Result<TOutput>
+  > {
+    if (!this.provider) {
+      throw new Error('Provider is required');
+    }
```

Comment:
This condition seems unnecessary since now the constructor requires a `walletOrProvider` to be non-null.

```suggestion
```